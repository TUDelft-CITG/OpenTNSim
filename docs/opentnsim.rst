=========
OpenTNSim
=========

This page lists all functions and classes available in the OpenTNSim.model and OpenTNSim.core modules. For examples on how to use these submodules please check out the Examples page, information on installing OpenTNim can be found on the Installation page.

Submodules
----------

The main components are the Model module and the Core module. All of their components are listed below. 

opentnsim\.model module
--------------------------

.. automodule:: opentnsim.model
   :members:
   :undoc-members:
   :show-inheritance:

opentnsim\.core module
--------------------------

.. automodule:: opentnsim.core
   :members:
   :undoc-members:
   :show-inheritance:

Module contents
---------------

.. automodule:: opentnsim
   :members:
   :undoc-members:
   :show-inheritance:
