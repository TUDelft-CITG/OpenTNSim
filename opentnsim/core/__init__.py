from .resource import HasResource
from .misc import Neighbours, ExtraMetadata
from .vessel_properties import HasLength, HasLoad, VesselProperties
from .container import HasContainer
from .movable import Movable, ContainerDependentMovable, Routable, Routeable
