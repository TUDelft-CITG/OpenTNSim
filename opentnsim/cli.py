# -*- coding: utf-8 -*-

"""Console script for opentnsim."""
# package(s) related to the command line interface
import sys

import click

import opentnsim.server


@click.group()
def cli(args=None):
    """OpenTNSim simulation."""
    click.echo("Replace this message by putting your code into " "opentnsim.cli.main")
    click.echo("See click documentation at http://click.pocoo.org/")
    return 0


@cli.command()
@click.option("--host", default="0.0.0.0")
@click.option("--port", default=5000, type=int)
@click.option("--debug/--no-debug", default=False)
def serve(host, port, debug, args=None):
    """Run a flask server with the backend code"""
    app = opentnsim.server.app
    app.run(host=host, debug=debug, port=port)


if __name__ == "__main__":
    sys.exit(cli())  # pragma: no cover
