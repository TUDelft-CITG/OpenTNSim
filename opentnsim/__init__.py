# -*- coding: utf-8 -*-

import opentnsim.core as core # now contains mixin class consumes energy, also the modified vessel properties
import opentnsim.graph_module as graph_module
import opentnsim.plot as plot
import opentnsim.model as model
import opentnsim.energy as energy
import opentnsim.strategy as strategy

"""Top-level package for OpenTNSim."""

__author__ = """Mark van Koningsveld"""
__email__ = "M.vanKoningsveld@tudelft.nl"
__version__ = "1.1.2"
