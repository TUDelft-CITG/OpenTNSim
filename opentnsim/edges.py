import opentnsim.core


def pass_berth(env, vessel, edge, anchorage_info):
    """Pass an edge that contains a 'Berth' with a vessel. A Berth can be a HasContainer (Quay) or a HasResource (Jetty)."""
    assert (
        "unload_duration" in vessel.metadata
    ), "If you want to unload at a quay a ship should have an unload duration"
    berth = edge["Berth"]

    # For the last edge:
    vessel.log_entry(
        "Start unloading at Berth",
        env.now,
        {"vessel": vessel, "berth": berth},
        berth.geometry,
    )
    # HasContainer -> Quay
    if isinstance(berth, opentnsim.core.HasContainer):
        berth.log_entry(
            "Start unloading at Berth",
            env.now,
            {"vessel": vessel, "berth": berth, "level": berth.container.level},
            berth.geometry,
        )
    # HasResource -> Jetty
    elif isinstance(berth, opentnsim.core.HasResource):
        berth.log_entry(
            "Start unloading at Berth",
            env.now,
            {"vessel": vessel, "berth": berth, "count": berth.resource.count},
            berth.geometry,
        )

    yield vessel.env.timeout(vessel.metadata["unload_duration"])
    vessel.log_entry(
        "Stop unloading at Berth",
        env.now,
        {"vessel": vessel, "berth": berth},
        vessel.geometry,
    )

    if isinstance(berth, opentnsim.core.HasContainer):
        berth.log_entry(
            "Stop unloading at Berth",
            env.now,
            {"vessel": vessel, "berth": berth, "level": berth.container.level},
            berth.geometry,
        )
        # give back space on the Quay
        yield berth.container.put(vessel.length)
        berth.log_entry(
            "Stop release reservation at Berth",
            env.now,
            {"vessel": vessel, "berth": berth, "level": berth.container.level},
            berth.geometry,
        )
    elif isinstance(berth, opentnsim.core.HasResource):
        berth.log_entry(
            "Stop unloading at Berth",
            env.now,
            {"vessel": vessel, "berth": berth, "count": berth.resource.count},
            berth.geometry,
        )
        # release the Jetty
        # we should have requested this spot in the anchorage area, now we can release it.
        berth_req = anchorage_info.get("berth_req")
        assert (
            berth_req is not None
        ), f"Arrived at berth of type resource with berth request {berth_req}\n for vessel \n{vessel} at edge:\n{edge}, with info \n{anchorage_info}"
        yield berth.resource.release(berth_req)
        berth.log_entry(
            "Stop release reservation at Berth",
            env.now,
            {"vessel": vessel, "berth": berth, "count": berth.resource.count},
            berth.geometry,
        )


def pass_anchorage(env, vessel, edge, route, graph, info):
    """Pass an ege with an anchorage. The route argument should be the remaining part of the voyage.
    The info argument should be a dictionary in which we can return some information on requested resources.
    This pass by reference is needed because this is a generator, which means we can not return something and we don't want to change objects inline.

    """
    assert (
        len(route) > 2
    ), "If there is an anchorage in the route, the route should be at least 2 nodes long"

    # TODO: This function is now coupled to berth behaviour.
    # We might want to extract the finding the next berth part and the waiting for an available berth spot by using a callback or something.

    # return a dictionary with information on the requests made in the anchorage area
    next_e = None

    # Some may ships make multiple stops
    # Traverse the route for the first location where you would want to stop
    # Find the next Berth
    for j in range(0, len(route) - 1):
        next_e = (route[j], route[j + 1])
        if "Berth" in graph.edges[next_e]:
            # found a next edge
            break
    else:
        raise ValueError(f"No berth found on route past anchorage {edge}")

    # This is the next stop
    next_edge = graph.edges[next_e]

    # we need to pass this along
    info["next_edge"] = next_edge
    info["out_of_system"] = False

    anchorage = edge["Anchorage"]

    # If there is no space, move out of the system
    if anchorage.resource.count >= anchorage.resource.capacity:
        info["out_of_system"] = True
        vessel.log_entry("Moving out of the system", env.now, "", vessel.geometry)
        return

    # The ship can now request a spot at the anchorage area
    vessel.log_entry("Start wait at Anchorage", env.now, "", vessel.geometry)
    anchorage.log_entry(
        "Request spot at Anchorage",
        env.now,
        {"berth": anchorage, "vessel": vessel, "count": anchorage.resource.count},
        anchorage.geometry,
    )
    # Using this context we will release this request at the end of the with statement
    with anchorage.resource.request() as anchor_req:
        # yielding this will advance time until we have our space at the anchorage
        yield anchor_req
        # Now we are at the anchorage we can request a spot at our target location
        # We know about two types of berths, the one with a container (think quay) an with a resource (thinkjetty)
        assert isinstance(
            next_edge["Berth"],
            (opentnsim.core.HasContainer, opentnsim.core.HasResource),
        )
        # Lookup the berth
        berth = next_edge["Berth"]
        vessel.log_entry(
            "Request spot at Berth",
            env.now,
            {"berth": berth, "vessel": vessel, "length": vessel.length},
            vessel.geometry,
        )
        if isinstance(berth, opentnsim.core.HasContainer):
            # We need information from the ship to be able to moore on a quay
            assert isinstance(
                vessel, opentnsim.core.VesselProperties
            ), "If you have quays (Berths with a container) in your network, all ships should have a length implemented with VesselPropertis"
            yield berth.container.get(vessel.length)
            berth.log_entry(
                "Start assign reservation at Berth",
                env.now,
                {"berth": berth, "vessel": vessel, "level": berth.container.level},
                berth.geometry,
            )
        elif isinstance(berth, opentnsim.core.HasResource):
            berth_req = berth.resource.request()
            # we need to release this request if we are done loading
            info["berth_req"] = berth_req
            yield berth_req
            berth.log_entry(
                "Start assign reservation at Berth",
                env.now,
                {"berth": berth, "vessel": vessel, "count": berth.resource.count},
                berth.geometry,
            )

    anchorage.log_entry(
        "Release spot at Anchorage",
        env.now,
        {"berth": anchorage, "vessel": vessel, "count": anchorage.resource.count},
        anchorage.geometry,
    )
    vessel.log_entry("Leaving Anchorage", env.now, "", vessel.geometry)
