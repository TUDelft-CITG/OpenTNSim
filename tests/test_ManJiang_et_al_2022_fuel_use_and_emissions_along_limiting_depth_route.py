'''Here we test the code for estimating fuel consumption and emission rates of CO2, PM10 and NOx for the three waterway sections along the route.'''


# Importing libraries

# package(s) related to time, space and id
import datetime, time

# you need these dependencies (you can get these from anaconda)
# package(s) related to the simulation
import simpy
import pandas as pd

# spatial libraries
import shapely.geometry

# package(s) for data handling
import numpy as np

# OpenTNSim
import opentnsim

# Used for making the graph to visualize our problem
import networkx as nx

import pytest

# Creating the test objects

# Actual testing starts here
# - tests 3 fixed velocities to return the right P_tot
# - tests 3 fixed power to return indeed the same P_tot
# - tests 3 fixed power to return indeed the same v
# todo: current tests do work with vessel.h_squat=True ... issues still for False
def test_simulation():

    Node = type("Site", (opentnsim.core.Identifiable, opentnsim.core.Locatable), {})

    data_node_1 = {"name": "Node 1", "geometry": shapely.geometry.Point(0, 0)}
    data_node_2 = {"name": "Node 2", "geometry": shapely.geometry.Point(0.8983, 0)}
    data_node_3 = {"name": "Node 3", "geometry": shapely.geometry.Point(1.7966, 0)}
    data_node_4 = {"name": "Node 4", "geometry": shapely.geometry.Point(2.6949, 0)}

    node_1 = Node(**data_node_1)
    node_2 = Node(**data_node_2)
    node_3 = Node(**data_node_3)
    node_4 = Node(**data_node_4)

    nodes = [node_1, node_2, node_3, node_4]

    """create a graph with 3 edges"""
    FG = nx.DiGraph()

    positions = {}
    for node in nodes:
        positions[node.name] = (node.geometry.x, node.geometry.y)
        FG.add_node(node.name, geometry=node.geometry)

    path = [
        [nodes[0], nodes[1]],  # From node 1 to node 2
        [nodes[1], nodes[2]],  # From node 2 to node 3
        [nodes[2], nodes[3]],  # From node 3 to node 4
        [nodes[3], nodes[2]],  # From node 4 to node 3
        [nodes[2], nodes[1]],  # From node 3 to node 2
        [nodes[1], nodes[0]],  # From node 2 to node 1
    ]

    for edge in path:
        # add depth to Info.GeneralDepth
        FG.add_edge(edge[0].name, edge[1].name, weight=1, Info={"GeneralDepth": 6})
    middle_edges = [
        (node_2.name, node_3.name),
        (node_3.name, node_2.name)
    ]

    for e in middle_edges:
        edge = FG.edges[e]
        edge['Info']['GeneralDepth'] = 2.5
        
    # Make your preferred class out of available mix-ins.
    TransportResource = type(
        "Vessel",
        (
            opentnsim.core.Identifiable,
            opentnsim.core.Movable,
            opentnsim.core.Routeable,
            opentnsim.core.VesselProperties,
            opentnsim.core.ConsumesEnergy,
            opentnsim.core.ExtraMetadata,
        ),
        {},
    )

    # Create a dict with all important settings
    data_vessel = {"env": None,
                   "name": None,
                   "route": None,
                   "geometry": None,
                   "v": None,  # m/s
                   "type": None,
                   "B": 11.4,
                   "L": 110,
                   "H_e": None,
                   "H_f": None,
                   "T": 2,
                   "safety_margin": 0.3,  # for tanker vessel with rocky bed the safety margin is recommended as 0.3 m
                   "h_squat": True,  # if consider the ship squatting while moving, set to True, otherwise set to False
                   "P_installed": 1750.0, # kW
                   "P_tot_given": None,  # kW
                   "bulbous_bow": False,  # if a vessel has no bulbous_bow, set to False; otherwise set to True.
                   "P_hotel_perc": 0.05,
                   "P_hotel": None,  # None: calculate P_hotel from percentage
                   "L_w": 3.0,
                   "C_B": 0.85,
                   "C_year": 1990,
                   }

    vessel = TransportResource(**data_vessel)

    path = nx.dijkstra_path(FG, nodes[0].name, nodes[3].name)

    # Actual testing starts here
    def run_simulation(V_s, P_tot_given):
        # Start simpy environment
        simulation_start = datetime.datetime.now()
        env = simpy.Environment(initial_time=time.mktime(simulation_start.timetuple()))
        env.epoch = time.mktime(simulation_start.timetuple())

        # Add graph to environment
        env.FG = FG

        # Add environment and path to the vessel
        # create a fresh instance of vessel
        vessel = TransportResource(**data_vessel)
        vessel.name = 'Vessel No.1'
        vessel.env = env  # the created environment
        vessel.route = path  # the route (the sequence of nodes, as stored as the second column in the path)
        vessel.geometry = env.FG.nodes[path[0]]['geometry']  # a shapely.geometry.Point(lon,lat) (here taken as the starting node of the vessel)
        vessel.v = V_s
        vessel.P_tot_given = P_tot_given

        # Start the simulation
        env.process(vessel.move())
        env.run()

        return vessel

    # prepare input data to loop through
    input_data = {'V_s': [3.0, 3.5, None, None], 'P_tot_given': [None, None, 212, 289]}

    # create empty plot data
    plot_data = {}

    # loop through the various input data
    for index, value in enumerate(input_data['V_s']):
        # Run a basic simulation with V_s and P_tot_given combi
        vessel = run_simulation(input_data['V_s'][index], input_data['P_tot_given'][index])

        # create an EnergyCalculation object and perform energy consumption calculation
        energycalculation = opentnsim.energy.EnergyCalculation(FG, vessel)
        energycalculation.calculate_energy_consumption()

        # create dataframe from energy calculation computation
        df = pd.DataFrame.from_dict(energycalculation.energy_use)

        # add/modify some comlums to suit our plotting needs
        df['fuel_kg_per_km'] = (df['total_fuel_consumption'] / 1000) / (df['distance'] / 1000)
        df['CO2_g_per_km'] = (df['total_emission_CO2']) / (df['distance'] / 1000)
        df['PM10_g_per_km'] = (df['total_emission_PM10']) / (df['distance'] / 1000)
        df['NOx_g_per_km'] = (df['total_emission_NOX']) / (df['distance'] / 1000)

        label = 'V_s = ' + str(input_data['V_s'][index]) + ' P_tot_given = ' + str(input_data['P_tot_given'][index])

        # Note that we make a dict to collect all plot data.
        # We use labels like ['V_s = None P_tot_given = 274 fuel_kg_km'] to organise the data in the dict
        # The [0, 0, 1, 1, 2, 2] below creates a list per section
        plot_data[label + ' fuel_kg_per_km']   = list(df.fuel_kg_per_km[[0, 0, 1, 1, 2, 2]])
        plot_data[label + ' CO2_g_per_km'] = list(df.CO2_g_per_km[[0, 0, 1, 1, 2, 2]])
        plot_data[label + ' PM10_g_per_km'] = list(df.PM10_g_per_km[[0, 0, 1, 1, 2, 2]])
        plot_data[label + ' NOx_g_per_km'] = list(df.NOx_g_per_km[[0, 0, 1, 1, 2, 2]])

    # todo: this test should be modified to test the fuel use and emission (looking at test name)
    # test the estimation of fuel consumption and emission rates of CO2, PM10 and NOx in section 1

# test the estimation of fuel consumption and emission rates of CO2, PM10 and NOx in section 1
    np.testing.assert_almost_equal(5.1776, plot_data['V_s = 3.0 P_tot_given = None fuel_kg_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(5.9119, plot_data['V_s = 3.5 P_tot_given = None fuel_kg_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
  
    np.testing.assert_almost_equal(16427.1882, plot_data['V_s = 3.0 P_tot_given = None CO2_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(18757.0696, plot_data['V_s = 3.5 P_tot_given = None CO2_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(11.7777, plot_data['V_s = 3.0 P_tot_given = None PM10_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(11.7543, plot_data['V_s = 3.5 P_tot_given = None PM10_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(251.6599, plot_data['V_s = 3.0 P_tot_given = None NOx_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(266.2670, plot_data['V_s = 3.5 P_tot_given = None NOx_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(5.1747, plot_data['V_s = None P_tot_given = 212 fuel_kg_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(5.9101, plot_data['V_s = None P_tot_given = 289 fuel_kg_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
   
    np.testing.assert_almost_equal(16418.0953, plot_data['V_s = None P_tot_given = 212 CO2_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(18751.2460, plot_data['V_s = None P_tot_given = 289 CO2_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(11.7786, plot_data['V_s = None P_tot_given = 212 PM10_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(11.7525, plot_data['V_s = None P_tot_given = 289 PM10_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(251.6164, plot_data['V_s = None P_tot_given = 212 NOx_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(266.2045, plot_data['V_s = None P_tot_given = 289 NOx_g_per_km'][0], decimal=2, err_msg='not almost equal', verbose=True) 
    
    # test the estimation of fuel consumption and emission rates of CO2, PM10 and NOx in section 2

    np.testing.assert_almost_equal(6.8960, plot_data['V_s = 3.0 P_tot_given = None fuel_kg_per_km'][1], decimal=0, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(12.7998, plot_data['V_s = 3.5 P_tot_given = None fuel_kg_per_km'][1], decimal=0, err_msg='not almost equal', verbose=True)
    
    np.testing.assert_almost_equal(21879.3371, plot_data['V_s = 3.0 P_tot_given = None CO2_g_per_km'][1], decimal=0, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(40610.4602, plot_data['V_s = 3.5 P_tot_given = None CO2_g_per_km'][1], decimal=0, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(13.7115, plot_data['V_s = 3.0 P_tot_given = None PM10_g_per_km'][1], decimal=0, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(22.4056, plot_data['V_s = 3.5 P_tot_given = None PM10_g_per_km'][1], decimal=0, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(310.5964, plot_data['V_s = 3.0 P_tot_given = None NOx_g_per_km'][1], decimal=0, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(559.1833, plot_data['V_s = 3.5 P_tot_given = None NOx_g_per_km'][1], decimal=0, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(5.7584, plot_data['V_s = None P_tot_given = 212 fuel_kg_per_km'][1], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(6.8940, plot_data['V_s = None P_tot_given = 289 fuel_kg_per_km'][1], decimal=2, err_msg='not almost equal', verbose=True)
  
    np.testing.assert_almost_equal(18270.0350, plot_data['V_s = None P_tot_given = 212 CO2_g_per_km'][1], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(21872.9348, plot_data['V_s = None P_tot_given = 289 CO2_g_per_km'][1], decimal=2, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(13.1072, plot_data['V_s = None P_tot_given = 212 PM10_g_per_km'][1], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(13.7091, plot_data['V_s = None P_tot_given = 289 PM10_g_per_km'][1], decimal=2, err_msg='not almost equal', verbose=True)

    np.testing.assert_almost_equal(279.9984, plot_data['V_s = None P_tot_given = 212 NOx_g_per_km'][1], decimal=2, err_msg='not almost equal', verbose=True)
    np.testing.assert_almost_equal(310.5219, plot_data['V_s = None P_tot_given = 289 NOx_g_per_km'][1], decimal=2, err_msg='not almost equal', verbose=True)
    

    
