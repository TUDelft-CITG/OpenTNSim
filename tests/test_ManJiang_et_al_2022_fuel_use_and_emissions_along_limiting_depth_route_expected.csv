V_s = 3.0 P_tot_given = None fuel_kg_per_km,V_s = 3.0 P_tot_given = None CO2_g_per_km,V_s = 3.0 P_tot_given = None PM10_g_per_km,V_s = 3.0 P_tot_given = None NOx_g_per_km,V_s = 3.5 P_tot_given = None fuel_kg_per_km,V_s = 3.5 P_tot_given = None CO2_g_per_km,V_s = 3.5 P_tot_given = None PM10_g_per_km,V_s = 3.5 P_tot_given = None NOx_g_per_km,V_s = None P_tot_given = 276 fuel_kg_per_km,V_s = None P_tot_given = 276 CO2_g_per_km,V_s = None P_tot_given = 276 PM10_g_per_km,V_s = None P_tot_given = 276 NOx_g_per_km,V_s = None P_tot_given = 391 fuel_kg_per_km,V_s = None P_tot_given = 391 CO2_g_per_km,V_s = None P_tot_given = 391 PM10_g_per_km,V_s = None P_tot_given = 391 NOx_g_per_km
6.693317158102194,21236.06989252423,13.415066225758228,302.58836710397065,7.509247789706042,23824.79526006735,13.953810033547331,327.89145361133905,6.652495193966169,21106.55293358357,13.377378443829297,301.208002346249,7.665599830841233,24320.85764512355,14.139942163485955,333.6027178936944
6.693317158102194,21236.06989252423,13.415066225758228,302.58836710397065,7.509247789706042,23824.79526006735,13.953810033547331,327.89145361133905,6.652495193966169,21106.55293358357,13.377378443829297,301.208002346249,7.665599830841233,24320.85764512355,14.139942163485955,333.6027178936944
10.451373972405932,33159.35923972427,18.76341373781608,449.98018561901796,13.693008192962896,43444.180539491375,23.957832635505433,601.6584713005019,8.173401577531651,25931.97409598678,16.435740602505,370.0707602625313,9.621811839925249,30527.384837581016,17.748364901549497,418.7359961858251
10.451373972405932,33159.35923972427,18.76341373781608,449.98018561901796,13.693008192962896,43444.180539491375,23.957832635505433,601.6584713005019,8.173401577531651,25931.97409598678,16.435740602505,370.0707602625313,9.621811839925249,30527.384837581016,17.748364901549497,418.7359961858251
6.693317158102193,21236.069892524225,13.415066225758228,302.58836710397065,7.509247789366389,23824.79525898973,13.953810033136598,327.89145359886385,6.652495193966172,21106.55293358358,13.3773784438293,301.20800234624903,7.665599830841235,24320.857645123557,14.139942163485955,333.60271789369443
6.693317158102193,21236.069892524225,13.415066225758228,302.58836710397065,7.509247789366389,23824.79525898973,13.953810033136598,327.89145359886385,6.652495193966172,21106.55293358358,13.3773784438293,301.20800234624903,7.665599830841235,24320.857645123557,14.139942163485955,333.60271789369443
